util/tls/tls_socket.cc

```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
+        // Falling through to read would likely result in a stall/deadlock.
+        LOG(DFATAL) << "SSL BUG: Engine returned NEED_WRITE but OutputPending is 0";
+        return make_unexpected(make_error_code(errc::protocol_error));
+      }
+
+      if (output_pending_bytes > 0) {
+        auto output_buf{engine_->PeekOutputBuf()};
+        auto send_result{next_sock_->TrySend(output_buf)};
+        if (!send_result) {
+          // Write failed (EAGAIN or Error).
+          if (total_bytes_read > 0)
+            break;
+          return make_unexpected(send_result.error());
+        }
+        engine_->ConsumeOutputBuf(*send_result);
+
+        // If we couldn't send everything, we can't proceed safely.
+        if ((*send_result) < output_buf.size()) {
+          if (total_bytes_read > 0)
+            break;
+          return make_unexpected(make_error_code(errc::resource_unavailable_try_again));
+        }
+
+        // If we handled a NEED_WRITE successfully, we loop again to see what engine wants next
+        if (read_result == Engine::NEED_WRITE)
+          continue;
+      }
+      ///////////////////////////////////////////////////////////////
+      // 3. Check for read conflict:
+      // A read conflict implies the application is polling TryRecv while concurrently
+      // blocked on Recv, which is a usage error. We check this to prevent buffer corruption/crash.
+      if (read_in_progress) {
+        if (total_bytes_read > 0)
+          break;
+        return make_unexpected(make_error_code(errc::resource_unavailable_try_again));
+      }
+      ///////////////////////////////////////////////////////////////
+      // 4. Handle Pending Reads
+      auto input_buf{engine_->PeekInputBuf()};
+      auto recv_result{next_sock_->TryRecv(input_buf)};
+      if (recv_result) {
+        if ((*recv_result) > 0) {
+          engine_->CommitInput(*recv_result);
+          // Loop back to call engine_->Read() again and decrypt these bytes.
+          continue;
+        }
+        // *recv_result == 0 (Upstream socket EOF)
+        // The engine returned NEED_READ, but the socket is "dead".
+        // This is a "Dirty EOF" (TCP closed without TLS close_notify).
+        // We must return connection_reset to signal the dirty shutdown.
+        if (total_bytes_read > 0)
+          break;
+
+        // If we are here: total_bytes_read==0 (no decrypted data to return), and socket is "dead".
+        // We report this as a connection reset/abort because it wasn't a clean TLS shutdown.
+        return make_unexpected(make_error_code(errc::connection_reset));
+      }
+
+      // recv_result has an error (e.g. EAGAIN)
+      if (total_bytes_read > 0)
+        break;
+      return make_unexpected(recv_result.error());
+    } else if (read_result == Engine::EOF_STREAM) {  // case 3: Abrupt EOF
+      if (total_bytes_read > 0) {
+        break;
+      }
+      return make_unexpected(make_error_code(errc::connection_reset));
```

Comment:
why it is abrupt eof? 
we return recv return 0 in case the stream ended. similarly TryRecv should return 0 here.